runtime/tr.source/trj9/ilgen/Walker.cpp

```diff
@@ -2188,13 +2188,20 @@ TR_J9ByteCodeIlGenerator::genTreeTop(TR::Node * n)
             _couldOSRAtNextBC = true;
             if ((n->getOpCode().isCheck() || n->getOpCodeValue() == TR::treetop) && n->getFirstChild()->getOpCode().isCall())
                {
-               // The current stack should contain the current method's state prior to the call,
-               // with the call's arguments popped off. If this call is later inlined, it may
-               // contain an OSR transition. Therefore, it is necessary to save the current
-               // method's state now, so that it is available when the transition is made.
-               handlePendingPushSaveSideEffects(n);
-               saveStack(-1);
-               stashPendingPushLivenessForOSR();
+               // Calls are only OSR points for their first evaluation. Other references to the call,
+               // that are also anchored under checks or treetops, are not OSR points. This can be
+               // identified based on its reference count.
+               //
+               if (n->getFirstChild()->getReferenceCount() == 1)
```

Comment:
So the logic here is that we are generating IL and so when we generate code for the call node tree top, it must have a reference count of 1 "currently" even if it's reference count could become more than 1 "eventually", correct ?

In which case I'm a bit worried if we are 100% sure that there is not be some situation in which a second reference to the call node could have been created before the treetop node for the call node was created; a secondary worry is that in the event of such a reference having been innocently created it would result in a functional bug (missing pending push stores).

Reply:
This is a valid concern. Calls must be placed under their own tree, as they have side-effects, however, there is nothing to guarantee the call's tree is processed by `genTreeTop` before being referenced under another node.

I cannot find a point where `genInvoke` or one of its callers are managing several nodes that could potentially reference the call. Moreover, in testing, only `checkcast` has failed this check so far, which is the desired effect.

I can understand if the evidence above is not sufficient, or if this is not a restriction on IlGen we wish to make. Another potential approach might be to use a checklist on the node index or their BCI, but this would incur additional overhead.

Reply:
I'm not concerned about the overhead of a checklist solution if it results in a more robust solution. So I would suggest exploring that option at least to see what it looks like rather than commit this change as it is.

Reply:
A stack allocated node checklist has been added. In testing, it behaves the same as the reference count check and should be more robust.

Reply:
Thanks this looks good to me now but it's probably good to get @andrewcraik to review again since it's changed since he looked at it last.